## Introduction

You've just learned how the dictionary data type uses key and value pairs to store information. And you've learned how you can use conditional logic to select, filter, or do different things to information based on their values. In this Milestone, you'll practice your skills with these topics by working with data from the NCAA! You'll answer some questions regarding universities with Division I American Football teams, their conferences, and enrollments.

Ready, set, hut! 

![](https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExZmY4N2IzNjAyMGQzNTQ4MDNlYmU1NDRlOTU0Y2Q1OWY4NzRjMTM5YyZjdD1n/A04VZmuyrLPYft1UyJ/giphy.gif)

To import this data, remember to run the following cell.

In [1]:
# load the data
import sys
sys.path.append('../files')

from ncaa_fbs import nickname, enrollment

Instead of printing each dictionary out, below we will provide code snippets of each dicitonary to give you an understanding of what they contain. If you decide you want to view these dictionaries yourself, just open up a new code cell and `print` the dictionaries to the screen!


### Nicknames

The `nickname` dictionary contains the name of the university (as keys) followed by its nickname (as values).

```python
{'Air Force': 'Falcons',
 'Akron': 'Zips',
 'Alabama': 'Crimson Tide',
 'Appalachian State': 'Mountaineers',
 'Arizona': 'Wildcats',
 'Arizona State': 'Sun Devils', #<-- Dr A's and Dr C's alma mater
 'Arkansas': 'Razorbacks',
 'Arkansas State': 'Red Wolves',
 'Army': 'Black Knights',
 'Auburn': 'Tigers',
 'Ball State': 'Cardinals',
 'Baylor': 'Bears',
 'Boise State': 'Broncos',
  ...
 'Western Michigan': 'Broncos',
 'Wisconsin': 'Badgers',
 'Wyoming': 'Cowboys'}
```

For example, the team from `Arizona State`, Dr. Crider's and Dr. Alvarez's alma mater, is also referred to as the `Sun Devils`. 

![](https://media.giphy.com/media/m9uzpnQvjFdORTzLDE/giphy.gif)


### Enrollments

The `enrollment` dictionary contains the number of students at each university.


```python
{'Air Force': 4181,
 'Akron': 14516,
 'Alabama': 38316,
 'Appalachian State': 20641,
 'Arizona': 49471,
 'Arizona State': 77881,
 'Arkansas': 29068,
 'Arkansas State': 12863,
 'Army': 4594,
 'Auburn': 31526,
 'Ball State': 19337,
 'Baylor': 20626,
 'Boise State': 25830,
  ...
 'Western Michigan': 19038,
 'Wisconsin': 47932,
 'Wyoming': 11479}
```

Let's preview the data for one university, `Purdue`, by passing this *key* to each one of the dictionaries and printing the associated *value*. Run the cell below.

In [2]:
# Preview the data for Purdue
print(nickname['Purdue'])
print(enrollment['Purdue'])

Boilermakers
49639


The nickname for Purdue is "Boilermakers" and have a student population of 49,639 students. Pretty simple once we have all that information stored in a dictionary!

## Tasks

### Task 1: How Many Teams?

![](https://media.giphy.com/media/B3HJTgnQZrMv0VpxnN/giphy-downsized-large.gif)

Let's start by getting an overview of the size of our data. How many universities are there in the dataset? Some context: there are a lot of universities in the United States with American Football teams, but the  data you will be working with consists of universities that play in the highest tier of college football, the Division I Football Bowl Subdivision (FBS). How many teams are there in the FBS? Store the answer in a variable named `n_teams`. 

<span style='background :#FFF59E' > <b>HINT</b>: Every dictionary provided contains <b>all</b> of the universities that compete in the NCAA Division I </span>

In [3]:
# How many FBS teams are there? 
# hint: what function tells you how many "things" a dictionary/list contains?

# complete the code for the n_teams variable 
n_teams = len(nickname)

# do not modify the line below
print(f'There are {n_teams} teams in the FBS.')

There are 131 teams in the FBS.


<span style='background :#DDD5F3' > If done correctly, your code should output:</span>

<code>There are 131 teams in the FBS.</code>

### Task 2: Team Nicknames

Which teams have the nickname "Tigers"? Use a loop to iterate through each team in the `nickname` dictionary. For each team, use a conditional to check `if` their nickname is "Tigers". If so, append the university name to a Python list named `matches`. 

Remember that to access the school's nickname, you have to pass the name of the school as the key!

In [4]:
# set up an empty list for matching universities
matches = []

# loop over universities and append to list of matches when a match is found
for university in nickname:
    if nickname[university] == 'Tigers':
        matches.append(university)


# print the required output
print(matches)

['Auburn', 'Clemson', 'LSU', 'Memphis', 'Missouri']


<span style='background :#DDD5F3'>If done correctly, your output should be a list containing <b>5</b> universities.</span>

### Task 3: University Enrollment

Let's explore how large these universities are, in terms of their enrollment. 

Use a loop to iterate through each school in the `enrollment` dictionary.
*   If a university's enrollment is smaller than 10,000, increase the counting variable named `small_unis` by `1`.
*   If a university's enrollment is larger than 50,000, increase the counting variable named `large_unis` by 1.

Are there more FBS universities with enrollments smaller than 10,000, or with enrollments larger than 50,000?

The code provided at the end will print your results to the screen. But you'll still need to answer this question using a markdown cell!

In [5]:
# set up lists for matching universities
small_unis = 0
large_unis = 0

# loop over all universities
for school in enrollment:
    num_students = enrollment[school]
    
    if num_students < 10000:
        small_unis += 1
    elif num_students > 50000:
        large_unis += 1

# print the required outputs
print(f'Smaller universities (< 10k enrolled): {small_unis}')
print(f'Largest universities (> 50k enrolled): {large_unis}')

Smaller universities (< 10k enrolled): 7
Largest universities (> 50k enrolled): 14


There are more FBS with enrollments larger than 50,000.

*(Double-click this cell to edit this text, and replace it with a sentence answering the question of Task 3!)*

In [6]:
# run this cell to import more data!
from ncaa_fbs import fbs_teams

### Task 4: Conference Enrollments

![](https://media.giphy.com/media/SWXQWWp4otbjpbJpYv/giphy.gif)

The dictionary variable used in this task is called `fbs_teams`. In it, each university is now associated with a list of values, instead of one dictionary for each value.

For each university, the associated list contains items in this order:
* Index 0: Nickname
* Index 1: City
* Index 2: State
* Index 3: Enrollment
* Index 4: Conference

We have provided you a print statement at the end for all three statistics (number of universities, total enrollment, average enrollment) to show in the cell output.

Here is a snippet of `fbs_teams` for a few entries in the dictionary:

```python
{ ...
 'Arizona State': ['Sun Devils', 'Tempe', 'Arizona', 77881, 'Pac-12'],
 'Arkansas': ['Razorbacks', 'Fayetteville', 'Arkansas', 29068, 'SEC'],
 'Arkansas State': ['Red Wolves', 'Jonesboro', 'Arkansas', 12863, 'Sun Belt'],
  ...
}
```


Just like before, let's preview the data for one university, `Arizona State`, by passing this *key* to the `fbs_teams` dictionary, and printing the associated *value*. Run the cell below.

In [7]:
# Run the cell, do not modify the code.

# preview the data for Arizona State

# returns the whole list of values
print(fbs_teams['Arizona State'])

# nickname
print(fbs_teams['Arizona State'][0])

# City
print(fbs_teams['Arizona State'][1])

# State
print(fbs_teams['Arizona State'][2])

# Enrollment
print(fbs_teams['Arizona State'][3])

# Conference
print(fbs_teams['Arizona State'][4])

['Sun Devils', 'Tempe', 'Arizona', 77881, 'Pac-12']
Sun Devils
Tempe
Arizona
77881
Pac-12


In the code above we can see that Arizona State are called the **Sun Devils**, are located in the city of **Tempe** in the state of **Arizona**. They have a total student population of **77,881**, and play in the **Pac-12** conference! 

We'll use this dictionary to answer the following questions: 
1. **Does the Big 12 conference actually have twelve schools?**  
2. **What is the total enrollment across these Big 12 universities?**
3. **What is the average enrollment across these Big 12 universities?**


One approach to answering these questions is to store the enrollment for each school in a `list`, called `students`, defined below. Then the provided print statement will use different functions to calculate the values needed. 


Although you won't need to code these results at the end of the following cell - it is already done for you - but it's important to remember the following:
- Each entry in `students` will correspond to a single university. 
- It's `len`gth will be the number of universities in the Big 12. 
- The `sum` will be the total enrolllment in the Big 12 conference. 
- The average enrollment can be calculated by combining the `len` and `sum` functions.

Use the code template below to complete Task 4.

In [28]:
# approach 1: using a list followed by summary functions
students = []

# loop over universities, and append enrollments to the list
for team in fbs_teams:
    if fbs_teams[team][4] == 'Big 12':
        enrollment = fbs_teams[team][3]
        students.append(enrollment)
    # ---------- YOUR CODE STARTS HERE ---------- #



    
    

    
    
    
    # ---------- YOUR CODE ENDS HERE ------------ #

# print the results
print(f'Number of Big 12 schools: {len(students)}')
print(f'Total students: {sum(students)}')
print(f'Average enrollment: {sum(students) / len(students)}')

Number of Big 12 schools: 10
Total students: 280990
Average enrollment: 28099.0


<span style='background :#DDD5F3'>If done correctly, you will have <b>10</b> schools in the Big 12 conference. </span>

A different approach is to use **accumulator** variables. An accumulator variable is a variable that is modified in each iteration of the loop. For example, if you are storing the number of students for each university, outside of the loop you will define `n_students = 0` but within the loop you will increment the value of `n_students` for every university that matches your criteria. Incrementing and assigning back to the same variable can be done with the `+=` operator.

In the code below, you have two accumulator variables, `n_students` and `n_universities` both set to zero intially. For each item in your loop that matches the criteria you will update the `n_students` variable based on the enrollment at that university and increment the `n_universities` by `1`.

The print statement provided at the end will show the results in the cell output.

In [32]:
# approach 2: using two numeric variables to accumulate the results
n_students = 0
n_universities = 0

# loop over universities, and accumulate values
for team in fbs_teams:
     if fbs_teams[team][4] == 'Big 12':
        n_students += fbs_teams[team][3]
        n_universities += 1

    
# print the results
print(f'Number of Big 12 schools: {n_universities}')
print(f'Total students: {n_students}')
print(f'Average enrollment: {n_students / n_universities}')

Number of Big 12 schools: 10
Total students: 280990
Average enrollment: 28099.0


Way to go! Time to do a touchdown dance!

![](https://media.giphy.com/media/NdrO9USb8683H1sVuo/giphy.gif)

## Level Up

Yet another way of storing the Division I universities information is by using a dictionary with *dictionaries* for values. In the task above you had a dictionary with a *list* for values. To get to the desired value, you would use the appropriate index. 

In the case of a dictionary with dictionary values, you would instead use the *key* associated with the desired value.

A snippet of the `fbs_teams_dict` would look like:

```python
{...
 'Arizona': {'nickname': 'Wildcats',
             'city': 'Tucson',
             'state': 'Arizona',
             'enrollment': 49471,
             'conference': 'Pac-12'},
 'Arizona State': {'nickname': 'Sun Devils',
                   'city': 'Tempe',
                   'state': 'Arizona',
                   'enrollment': 77881,
                   'conference': 'Pac-12'},
 'Arkansas': {'nickname': 'Razorbacks',
              'city': 'Fayetteville',
              'state': 'Arkansas',
              'enrollment': 29068,
              'conference': 'SEC'},
...
```



Compare this to the `fbs_teams` structure you used for Task 4. If we want to get the enrollment for Purdue University, we still have 'Purdue' as the outer dictionary key, but now we specify the 'enrollment' feature for the inner dictionary key.




In [33]:
# import data
from ncaa_fbs import fbs_teams_dict

In [34]:
# get all the values for Purdue (output is a dictionary!)
fbs_teams_dict['Purdue']

{'nickname': 'Boilermakers',
 'city': 'West Lafayette',
 'state': 'Indiana',
 'enrollment': 49639,
 'conference': 'Big Ten'}

Compare this to the `fbs_teams` structure you used for Task 4. If we want to get the enrollment for Purdue University, we still have 'Purdue' as the outer dictionary key, but now we specify the `'enrollment'` feature for the inner dictionary key:

In [35]:
# use the appropriate key to get the enrollment for Purdue University.
# recall that before we used the appropriate *index* to get the enrollment.
fbs_teams_dict['Purdue']['enrollment']

49639

<span style='background :#FFF59E'>If you're feeling up for the additional practice, try replicating the tasks in this Milestone assignment but only using the `fbs_teams_dict` dictionary!</span>